File: ui-tests/Utils/DebugLogger.cs
using System;
using System.IO;
using System.Threading;

namespace UiTests.Utils;

/// <summary>
/// Lightweight file logger for ad-hoc diagnostics during UI test debugging.
/// </summary>
public static class DebugLogger
{
    private static readonly object SyncRoot = new();
    private static string? _logFilePath;
    private static int _initialized;
    private static readonly AsyncLocal<bool?> ScopeEnabled = new();
    private static readonly bool DefaultEnabled = ResolveDefaultEnabled();

    private static string ResolveLogFilePath()
    {
        if (_logFilePath is not null)
        {
            return _logFilePath;
        }

        var path = Environment.GetEnvironmentVariable("UITESTS_DEBUG_LOG");
        if (string.IsNullOrWhiteSpace(path))
        {
            path = Path.Combine(AppContext.BaseDirectory, "ui-tests-debug.log");
        }

        var directory = Path.GetDirectoryName(path);
        if (!string.IsNullOrWhiteSpace(directory) && !Directory.Exists(directory))
        {
            Directory.CreateDirectory(directory);
        }

        _logFilePath = path;
        return _logFilePath;
    }

    /// <summary>
    /// Clears the current log file.
    /// </summary>
    public static void Reset()
    {
        if (!IsEnabled)
        {
            return;
        }

        lock (SyncRoot)
        {
            var path = ResolveLogFilePath();
            File.WriteAllText(path, string.Empty);
            Interlocked.Exchange(ref _initialized, 1);
        }
    }

    /// <summary>
    /// Appends a message to the debug log with a UTC timestamp.
    /// </summary>
    public static void Log(string message)
    {
        if (!IsEnabled)
        {
            return;
        }

        lock (SyncRoot)
        {
            var path = ResolveLogFilePath();
            if (Interlocked.CompareExchange(ref _initialized, 1, 0) == 0 && !File.Exists(path))
            {
                File.WriteAllText(path, string.Empty);
            }
            var line = $"{DateTimeOffset.UtcNow:O} {message}";
            File.AppendAllText(path, line + Environment.NewLine);
        }
    }

    /// <summary>
    /// Disables or enables logging within the current async context.
    /// </summary>
    public static IDisposable PushScope(bool enabled)
        => new LoggerScope(enabled);

    /// <summary>
    /// Returns true when logging is enabled for the current async context.
    /// </summary>
    public static bool IsEnabled => ScopeEnabled.Value ?? DefaultEnabled;

    private static bool ResolveDefaultEnabled()
    {
        var raw = Environment.GetEnvironmentVariable("UITESTS_DEBUG_LOG_DEFAULT");
        if (string.IsNullOrWhiteSpace(raw))
        {
            return false;
        }

        if (raw.Equals("1", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        return bool.TryParse(raw, out var parsed) && parsed;
    }

    private sealed class LoggerScope : IDisposable
    {
        private readonly bool? _previous;

        public LoggerScope(bool enabled)
        {
            _previous = ScopeEnabled.Value;
            ScopeEnabled.Value = enabled;
        }

        public void Dispose()
        {
            ScopeEnabled.Value = _previous;
        }
    }
}

File: src/tests/gui/lib/debug-logger.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as process from "node:process";

/**
 * File-based debug logger for UI tests.
 *
 * Controlled by environment variables:
 *   UITESTS_DEBUG_LOG       - Path to log file (enables logging)
 *   UITESTS_DEBUG_LOG_DEFAULT - "1" to enable with default path
 *
 * Port of ui-tests/Utils/DebugLogger.cs
 */
class DebugLogger {
  private _logPath: string | null = null;
  private _enabled = false;

  constructor() {
    const envPath = process.env.UITESTS_DEBUG_LOG;
    const envDefault = process.env.UITESTS_DEBUG_LOG_DEFAULT;

    if (envPath && envPath.length > 0) {
      this._logPath = envPath;
      this._enabled = true;
    } else if (envDefault === "1") {
      this._logPath = path.join(process.cwd(), "ui-tests-debug.log");
      this._enabled = true;
    }
  }

  get isEnabled(): boolean {
    return this._enabled;
  }

  set isEnabled(value: boolean) {
    this._enabled = value;
  }

  /**
   * Logs a timestamped message to the log file.
   * No-op if logging is not enabled.
   */
  log(message: string): void {
    if (!this._enabled || !this._logPath) {
      return;
    }

    const timestamp = new Date().toISOString();
    const line = `[${timestamp}] ${message}\n`;

    try {
      fs.appendFileSync(this._logPath, line, "utf-8");
    } catch {
      // Silently ignore write failures to avoid test interference.
    }
  }

  /**
   * Clears the log file.
   */
  reset(): void {
    if (!this._logPath) {
      return;
    }

    try {
      fs.writeFileSync(this._logPath, "", "utf-8");
    } catch {
      // Silently ignore.
    }
  }
}

export const debugLogger = new DebugLogger();
